cli/src/alluxio.org/cli/cmd/info/report.go

```diff
@@ -80,9 +84,134 @@ func (c *ReportCommand) Run(args []string) error {
 		}
 		reportArg = args[0]
 	}
-	// TODO: output all in a serializable format and filter/trim as specified by flags
-	if c.format != "json" && c.format != "yaml" {
-		return stacktrace.NewError("Invalid format %v, must be one of [json, yaml]", c.format)
+
+	buf := &bytes.Buffer{}
+	if err := c.RunWithIO([]string{reportArg}, nil, buf, os.Stderr); err != nil {
+		io.Copy(os.Stdout, buf)
+		return err
+	}
+
+	obj := orderedmap.New()
+	if err := json.Unmarshal(buf.Bytes(), &obj); err != nil {
+		return stacktrace.Propagate(err, "error unmarshalling json from java command")
+	}
+
+	if reportArg == "summary" {
```

Comment:
this approach is very brute force, but it is hard to determine which fields need to be formatted. i have 2 ideas:

1. represent the output as a struct. set annotations on struct fields to determine if any conversion function should be applied when serializing
  - pros: define exactly which fields need to be converted, no chance for mistakenly converting the wrong field
  - cons: need to define a struct for the output and the struct needs to be updated if the output is updated
2. define a series of patterns/rules to determine if a field needs to be converted (ex. if field name ends with `Bytes`, cast value to `float64` and run `convertBytesToString`
  - pros: no need to do anything if fields are changed
  - cons: need to be careful with field names. also it can be hard to navigate the output if the output is not a flat map (ex. some fields have a map as a value instead of a string or number)

example for 1:
```
type SummaryOutput struct {
  FreeCapacity float64 `json:"freeCapacity" yaml:"freeCapacity" format:"convertBytesToString"`
  Started float64 `json:"started" yaml:"started" format:"convertMsToDatetime"`
  ...
}

func (o SummaryOutput) MarshalJSON() ([]byte, error) {
  val := reflect.Value(o)
  // the below can be abstracted to a function, reusable for each struct
  obj := orderedmap.New()
  for i := 0; i < val.NumField(); i++ {
    f := val.Field(i)
    convertName, ok := f.Tag.Lookup("format")
    if !ok {
      obj.Set(<fieldName>, <fieldValue>)
      continue
    }
    convertFunc, ok := convertFuncMap[convertName] // convertFuncMap is a mapping of name (ex. "convertBytesToString") to the function definition
    // check ok
    obj.Set(<fieldName>, convertFunc(<fieldValue>)
  } 
  return json.Marshal(obj)
}

var summary SummaryOutput
json.Unmarshal(buf, &summary)
json.Marshal(summary)
```

example for 2:
use the argument suffix as the way to determine if the value should be formatted.
- if the key ends with `Bytes` -> cast value to float64 and run `convertBytesToString`
- if the key ends with `Date` -> cast value to float64 and run `convertMsToDatetime`
- if the key ends with `Duration` -> cast the value to float64 and run `convertMsToDuration`
- if the key ends with `Tiers` -> assume the value is a map of key -> bytes

this will require some keys to be renamed to fit the above mapping
- freeCapacity -> freeCapacityBytes
- started -> startDate
- uptime -> uptimeDuration

```
// after populating ordered map obj
out := orderedMap.New()
for k, v := obj {
  if strings.HasSuffix(k, "Bytes") {
    out.Set(k, convertBytesToString(v.(float64))
  else if ...
}
json.Marshal(out)
```

Reply:
I prefer example 2, it's right time to review all summary property formats and keep them in similar formats.
But there is one more thing, need to think about a way to handle nested JSON conversion.
There are three scenarios:

1. The key-value pair to convert is in an object node. e.g. `"report": {"keyBytes": valueBytes}`
2. The key-value pair to convert is in an array node. e.g. `"reports": [{"keyBytes": valueBytes}, {...}]`
3. The key-value pair to convert is in a map. e.g. `"report": {"keyMap": {"keyBytes": valueBytes}, ...}`

I might need to think about all scenarios.